packages/docusaurus-theme-mermaid/src/theme/Mermaid/index.tsx

```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {useEffect, useState, useRef, useCallback} from 'react';
+import mermaid from 'mermaid';
+import useMermaidTheme from '@theme/useMermaidTheme';
+import type {Props} from '@theme/Mermaid';
+
+export default function Mermaid({value}: Props): JSX.Element {
+  const [svg, setSvg] = useState('');
+  const theme = useMermaidTheme();
+  // Assign a unique ID to each mermaid svg as per requirements of
+  // `mermaid.render`.
+  const id = useRef(0);
+
+  const render = useCallback(() => {
+    mermaid.render(`mermaid-svg-${id.current}`, value, (renderedSvg) =>
+      setSvg(renderedSvg),
+    );
+    id.current += 1;
+  }, [value]);
+
+  useEffect(() => {
+    render();
+  }, [theme, render]);
```

Comment:
We could SSR, but I didn't, because it would produce FOUC if the user is in dark mode—the same reason for Prism code blocks. I think not showing anything is not too bad, considering regular images would have similar UX.

Reply:
Yes, I agree

Reply:
Technically we could do like our ThemedImage component and render in both light/dark mode and hide one, but that looks a bit expensive for little ROI 🤷‍♂️ 

Is Mermaid even compatible with SSR in the first place? 

Does it support concurrency (ie rendering multiple pages in //) or is it a big singleton that you can just initialize once?

Is the render callback executed synchronously?

I don't see why rendering an SVG on a Node.js process synchronously wouldn't be possible, but it looks like the Mermaid library doesn't make this easy by design, so it's probably not worth it.

Any example in the wild of server-side-rendered Mermaid diagrams? I can't find any after a quick search.

Reply:
I'm not a fan of this component, would rather clean it a bit before merging 😅